README
Welcome to JSONBOT
==================

JSONBOT is a remote event-driven framework for building bots that talk JSON to 
each other over XMPP. 

This distribution provides bots built on this framework for console, IRC,
XMPP, Convore and WWW on the shell.

the jsb pakage contains the following programs:

    * jsb - console version of jsb
    * jsb-convore - Convore version of jsb 
    * jsb-makecert - create keys and certifictes for web console SSL
    * jsb-init - create data directory and config examples, default ~/.jsb
    * jsb-irc - IRC version of jsb
    * jsb-fleet - mix IRC and XMPP bots
    * jsb-sed - sed a whole directory
    * jsb-stop - stop a running bot
    * jsb-tornado - a shell web server based on tornado
    * jsb-udp - send udp packets to the bot that will relay the data
    * jsb-udpdirect - a stand alone udp program, copy & edit
    * jsb-xmpp - XMPP version of jsb

note: JSONBOT is in BETA stage right now and still subject to change of protocols and API.

see http://jsonbot.googlecode.com.
see https://jsonbot.org for documentation on the bot. 

if you want the GAE part of the bot install from mercurial repo.

* "hg clone http://jsonbot.googlecode.com/hg mybot"

license
~~~~~~~

JSONBOT is free code (MIT) and can be cloned where needed. 

contact the developer
~~~~~~~~~~~~~~~~~~~~~

    * email: bthate@gmail.com
    * jabber/xmpp: bthate@gmail.com
    * IRC: botfather on #dunkbots irc.freenode.net
    * twitter: http://twitter.com/jsonbot

urls
~~~~

    * Source code: http://jsonbot.googlecode.com
    * web: http://jsonbot.org:10102 (web)
    * jabber: jsonbot@jsonbot.org

running a development version of the bot
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

first checkout the main bot from the mercurial repository

    hg clone http://jsonbot.googlecode.com/hg mybot

now you can run the programs in the bin directory with the
./bin/<program>  command. try ./bin/jsb for the console app



Changelog

0.84.4
~~~~~~


3144 - Mar 23 17:54:06 2012 - summary:     Added tag 0.84.4 for changeset e7a3b8e53abe

3143 - Mar 23 17:35:55 2012 - summary:     fix

3142 - Mar 23 17:30:55 2012 - summary:     catch missing user Exception

3141 - Mar 23 17:19:40 2012 - summary:     setdefault values of fleet config vars

3140 - Mar 23 16:56:28 2012 - summary:     more fixes

3139 - Mar 23 16:47:13 2012 - summary:     bunch of crap .. getting tired fixing bugs that keep reoccuring

3138 - Mar 23 12:29:15 2012 - summary:     this is 0.84.4

3137 - Mar 23 11:21:05 2012 - summary:     fix bugs discovered by ./bin/jsb-makedocs

3136 - Mar 23 10:49:13 2012 - summary:     add release doc

3135 - Mar 23 08:37:13 2012 - summary:     upper loglevel

3134 - Mar 23 01:31:53 2012 - summary:     lower loglevel .. default globalcc to ""

3133 - Mar 23 01:25:04 2012 - summary:     first check for channel cc then global cc otherwiser set to "!;"

3132 - Mar 22 18:39:53 2012 - summary:     check if bot exists

3131 - Mar 22 18:15:25 2012 - summary:     check if bot exists

3130 - Mar 22 01:16:39 2012 - summary:     this is 0.84.3

3129 - Mar 22 00:59:54 2012 - summary:     lower loglevel

3128 - Mar 21 21:02:15 2012 - summary:     fix makechangelog


0.84.2
~~~~~~


3127 - Mar 21 21:02:03 2012 - summary:     Added tag 0.84.2 for changeset 49b68ccb6ab3

3126 - Mar 19 21:04:42 2012 - summary:     fix botlogs

3125 - Mar 19 14:31:19 2012 - summary:     enable the keepalive thread

3124 - Mar 19 13:55:23 2012 - summary:     fix reconnect code (IRC)

3123 - Mar 18 19:39:00 2012 - summary:     fix reconnect part 1

3122 - Mar 18 15:45:32 2012 - summary:     lower RequireError to loglevel info

3121 - Mar 18 15:19:08 2012 - summary:     tune configuration saving (include other possible options)

3120 - Mar 18 14:47:08 2012 - summary:     wait for connectok

3119 - Mar 18 14:44:30 2012 - summary:     lower loglevel

3118 - Mar 18 14:25:32 2012 - summary:     add possible config vars

3117 - Mar 18 14:04:53 2012 - summary:     dont reconnect on failed start

3116 - Mar 18 12:52:22 2012 - summary:     tune the fleet

3115 - Mar 17 19:41:21 2012 - summary:     fix setup.py

3114 - Mar 17 19:37:39 2012 - summary:     remove un needed packages from contrib

3113 - Mar 17 18:57:02 2012 - summary:     tune

3112 - Mar 17 16:06:05 2012 - summary:     logging nits

3111 - Mar 17 15:44:58 2012 - summary:     check for RequireError

3110 - Mar 17 15:38:52 2012 - summary:     use opts.nocolor

3109 - Mar 17 15:25:34 2012 - summary:     tune loglevel .. "improve" logging format

3108 - Mar 17 02:00:38 2012 - summary:     remove convore bot from the distro

3107 - Mar 17 01:05:03 2012 - summary:     fix typo

3106 - Mar 17 01:03:45 2012 - summary:     check for RequireError

3105 - Mar 17 00:54:01 2012 - summary:     check for RequireError

3104 - Mar 17 00:51:20 2012 - summary:     check for RequireError

3103 - Mar 17 00:46:13 2012 - summary:     dont forget to import

3102 - Mar 17 00:45:52 2012 - summary:     catch RequireError in EventBase.execute()

3101 - Mar 17 00:31:43 2012 - summary:     properly import RequireError

3100 - Mar 17 00:27:05 2012 - summary:     lock the getfleet

3099 - Mar 17 00:25:19 2012 - summary:     load the fleet bots in getfleet

3098 - Mar 17 00:05:18 2012 - summary:     add example to fish docs

3097 - Mar 16 23:52:51 2012 - summary:     put txt inside exception

3096 - Mar 16 23:37:34 2012 - summary:     fix

3095 - Mar 16 23:35:28 2012 - summary:     its 2012 now

3094 - Mar 16 23:34:30 2012 - summary:     fix typo

3093 - Mar 16 23:07:23 2012 - summary:     disable rss-clone commands

3092 - Mar 16 23:00:14 2012 - summary:     update docs

3091 - Mar 16 21:58:01 2012 - summary:     add config know to fish plugin

3090 - Mar 16 21:35:14 2012 - summary:     Corrected a typo in jsb.lib.errors

3089 - Mar 16 21:27:36 2012 - summary:     implement a RequireError

3088 - Mar 16 19:36:19 2012 - summary:     add bot.cfg.onconnect .. string to send after bot has logged in

3087 - Mar 16 19:25:34 2012 - summary:     add onconnect and onconnectmode configuration strings

3086 - Mar 16 18:11:42 2012 - summary:     fix test-plugs

3085 - Mar 16 17:16:27 2012 - summary:     check if fleet is not empty

3084 - Mar 16 16:14:38 2012 - summary:     fix fleet-cmnd ... use a control char !

3083 - Mar 16 15:14:52 2012 - summary:     tune fleet bot registering

3082 - Mar 16 14:03:04 2012 - summary:     update fleet booting

3081 - Mar 16 10:41:15 2012 - summary:     add bot to fleet in BotBase.start()

3080 - Mar 15 14:50:26 2012 - summary:     use unicode string

3079 - Mar 15 07:44:20 2012 - summary:     dont add multiple tornado bots

3078 - Mar 15 07:41:42 2012 - summary:     fix done command

3077 - Mar 15 07:40:04 2012 - summary:     add fleet-save

3076 - Mar 15 07:18:55 2012 - summary:     fix Fleet.addbot()

3075 - Mar 15 06:14:11 2012 - summary:     make boot work on gae

3074 - Mar 15 03:37:17 2012 - summary:     properly load plugins (or not)

3073 - Mar 15 02:54:58 2012 - summary:     update API doc

3072 - Mar 15 02:50:27 2012 - summary:     update

3071 - Mar 15 02:31:05 2012 - summary:     update pluginlist

3070 - Mar 14 22:05:43 2012 - summary:     fix fleet.loadall

3069 - Mar 14 21:59:07 2012 - summary:     force loading of plugins if myplugs had changed

3068 - Mar 14 21:47:07 2012 - summary:     remove dns from the contrib

3067 - Mar 14 21:33:53 2012 - summary:     fix sleek

3066 - Mar 14 19:42:45 2012 - summary:     detect config file changes on boot

3065 - Mar 14 18:27:35 2012 - summary:     provide a !qq command that quits the bot

3064 - Mar 14 13:24:53 2012 - summary:     show proper done() message

3063 - Mar 14 13:21:53 2012 - summary:     put loggin message at the right place

3062 - Mar 14 13:05:57 2012 - summary:     print message if db driver is disabled

3061 - Mar 14 04:42:31 2012 - summary:     set aliases

3060 - Mar 14 04:27:04 2012 - summary:     unload plugins when needed


files/toudp.py
#!/usr/bin/env python
#
#

""" send udp packets with data to be displayed to the jsb bot. """

## boot

import socket
import sys
import os

sys.path.insert(0, os.getcwd())

## commandline options

from optparse import OptionParser

parser = OptionParser(usage='usage: %prog [options]')

parser.add_option('-p', '--printto', type='string', default=False, dest='printto',
                  metavar='PATH',
                  help="channel/user to print to")
parser.add_option('-c', '--config', type='string', default="udp-send", dest='configname',
                  help="Specify a config file")
parser.add_option('-s', '--save', action='store_true', default=False, dest='saveconfig',
                  help="save to config file")

opts, args = parser.parse_args()
opts.args = args

host = 'localhost'
port = 5500
passwd = 'mekker'
printto = '#dunkbots'
seed = "blablablablablaz"
length = 16
ipv6 = 0


"""
A pure python (slow) implementation of rijndael with a decent interface

To include -

from rijndael import rijndael

To do a key setup -

r = rijndael(key, block_size = 16)

key must be a string of length 16, 24, or 32
blocksize must be 16, 24, or 32. Default is 16

To use -

ciphertext = r.encrypt(plaintext)
plaintext = r.decrypt(ciphertext)

If any strings are of the wrong length a ValueError is thrown
"""

# ported from the Java reference code by Bram Cohen, April 2001
# this code is public domain, unless someone makes
# an intellectual property claim against the reference
# code, in which case it can be made public domain by
# deleting all the comments and renaming all the variables

import copy
import string

shifts = [[[0, 0], [1, 3], [2, 2], [3, 1]],
          [[0, 0], [1, 5], [2, 4], [3, 3]],
          [[0, 0], [1, 7], [3, 5], [4, 4]]]

# [keysize][block_size]
num_rounds = {16: {16: 10, 24: 12, 32: 14}, 24: {16: 12, 24: 12, 32: 14}, 32: {16: 14, 24: 14, 32: 14}}

A = [[1, 1, 1, 1, 1, 0, 0, 0],
     [0, 1, 1, 1, 1, 1, 0, 0],
     [0, 0, 1, 1, 1, 1, 1, 0],
     [0, 0, 0, 1, 1, 1, 1, 1],
     [1, 0, 0, 0, 1, 1, 1, 1],
     [1, 1, 0, 0, 0, 1, 1, 1],
     [1, 1, 1, 0, 0, 0, 1, 1],
     [1, 1, 1, 1, 0, 0, 0, 1]]

# produce log and alog tables, needed for multiplying in the
# field GF(2^m) (generator = 3)
alog = [1]
for i in range(255):
    j = (alog[-1] << 1) ^ alog[-1]
    if j & 0x100 != 0:
        j ^= 0x11B
    alog.append(j)

log = [0] * 256
for i in range(1, 255):
    log[alog[i]] = i

# multiply two elements of GF(2^m)
def mul(a, b):
    if a == 0 or b == 0:
        return 0
    return alog[(log[a & 0xFF] + log[b & 0xFF]) % 255]

# substitution box based on F^{-1}(x)
box = [[0] * 8 for i in range(256)]
box[1][7] = 1
for i in range(2, 256):
    j = alog[255 - log[i]]
    for t in range(8):
        box[i][t] = (j >> (7 - t)) & 0x01

B = [0, 1, 1, 0, 0, 0, 1, 1]

# affine transform:  box[i] <- B + A*box[i]
cox = [[0] * 8 for i in range(256)]
for i in range(256):
    for t in range(8):
        cox[i][t] = B[t]
        for j in range(8):
            cox[i][t] ^= A[t][j] * box[i][j]

# S-boxes and inverse S-boxes
S =  [0] * 256
Si = [0] * 256
for i in range(256):
    S[i] = cox[i][0] << 7
    for t in range(1, 8):
        S[i] ^= cox[i][t] << (7-t)
    Si[S[i] & 0xFF] = i

# T-boxes
G = [[2, 1, 1, 3],
    [3, 2, 1, 1],
    [1, 3, 2, 1],
    [1, 1, 3, 2]]

AA = [[0] * 8 for i in range(4)]

for i in range(4):
    for j in range(4):
        AA[i][j] = G[i][j]
        AA[i][i+4] = 1

for i in range(4):
    pivot = AA[i][i]
    if pivot == 0:
        t = i + 1
        while AA[t][i] == 0 and t < 4:
            t += 1
            assert t != 4, 'G matrix must be invertible'
            for j in range(8):
                AA[i][j], AA[t][j] = AA[t][j], AA[i][j]
            pivot = AA[i][i]
    for j in range(8):
        if AA[i][j] != 0:
            AA[i][j] = alog[(255 + log[AA[i][j] & 0xFF] - log[pivot & 0xFF]) % 255]
    for t in range(4):
        if i != t:
            for j in range(i+1, 8):
                AA[t][j] ^= mul(AA[i][j], AA[t][i])
            AA[t][i] = 0

iG = [[0] * 4 for i in range(4)]

for i in range(4):
    for j in range(4):
        iG[i][j] = AA[i][j + 4]

def mul4(a, bs):
    if a == 0:
        return 0
    r = 0
    for b in bs:
        r <<= 8
        if b != 0:
            r = r | mul(a, b)
    return r

T1 = []
T2 = []
T3 = []
T4 = []
T5 = []
T6 = []
T7 = []
T8 = []
U1 = []
U2 = []
U3 = []
U4 = []

for t in range(256):
    s = S[t]
    T1.append(mul4(s, G[0]))
    T2.append(mul4(s, G[1]))
    T3.append(mul4(s, G[2]))
    T4.append(mul4(s, G[3]))

    s = Si[t]
    T5.append(mul4(s, iG[0]))
    T6.append(mul4(s, iG[1]))
    T7.append(mul4(s, iG[2]))
    T8.append(mul4(s, iG[3]))

    U1.append(mul4(t, iG[0]))
    U2.append(mul4(t, iG[1]))
    U3.append(mul4(t, iG[2]))
    U4.append(mul4(t, iG[3]))

# round constants
rcon = [1]
r = 1
for t in range(1, 30):
    r = mul(2, r)
    rcon.append(r)

del A
del AA
del pivot
del B
del G
del box
del log
del alog
del i
del j
del r
del s
del t
del mul
del mul4
del cox
del iG

class rijndael:
    def __init__(self, key, block_size = 16):
        if block_size != 16 and block_size != 24 and block_size != 32:
            raise ValueError('Invalid block size: ' + str(block_size))
        if len(key) != 16 and len(key) != 24 and len(key) != 32:
            raise ValueError('Invalid key size: ' + str(len(key)))
        self.block_size = block_size

        ROUNDS = num_rounds[len(key)][block_size]
        BC = block_size / 4
        # encryption round keys
        Ke = [[0] * BC for i in range(ROUNDS + 1)]
        # decryption round keys
        Kd = [[0] * BC for i in range(ROUNDS + 1)]
        ROUND_KEY_COUNT = (ROUNDS + 1) * BC
        KC = len(key) / 4

        # copy user material bytes into temporary ints
        tk = []
        for i in range(0, KC):
            tk.append((ord(key[i * 4]) << 24) | (ord(key[i * 4 + 1]) << 16) |
                (ord(key[i * 4 + 2]) << 8) | ord(key[i * 4 + 3]))

        # copy values into round key arrays
        t = 0
        j = 0
        while j < KC and t < ROUND_KEY_COUNT:
            Ke[t / BC][t % BC] = tk[j]
            Kd[ROUNDS - (t / BC)][t % BC] = tk[j]
            j += 1
            t += 1
        tt = 0
        rconpointer = 0
        while t < ROUND_KEY_COUNT:
            # extrapolate using phi (the round key evolution function)
            tt = tk[KC - 1]
            tk[0] ^= (S[(tt >> 16) & 0xFF] & 0xFF) << 24 ^  \
                     (S[(tt >>  8) & 0xFF] & 0xFF) << 16 ^  \
                     (S[ tt        & 0xFF] & 0xFF) <<  8 ^  \
                     (S[(tt >> 24) & 0xFF] & 0xFF)       ^  \
                     (rcon[rconpointer]    & 0xFF) << 24
            rconpointer += 1
            if KC != 8:
                for i in range(1, KC):
                    tk[i] ^= tk[i-1]
            else:
                for i in range(1, KC / 2):
                    tk[i] ^= tk[i-1]
                tt = tk[KC / 2 - 1]
                tk[KC / 2] ^= (S[ tt        & 0xFF] & 0xFF)       ^ \
                              (S[(tt >>  8) & 0xFF] & 0xFF) <<  8 ^ \
                              (S[(tt >> 16) & 0xFF] & 0xFF) << 16 ^ \
                              (S[(tt >> 24) & 0xFF] & 0xFF) << 24
                for i in range(KC / 2 + 1, KC):
                    tk[i] ^= tk[i-1]
            # copy values into round key arrays
            j = 0
            while j < KC and t < ROUND_KEY_COUNT:
                Ke[t / BC][t % BC] = tk[j]
                Kd[ROUNDS - (t / BC)][t % BC] = tk[j]
                j += 1
                t += 1
        # inverse MixColumn where needed
        for r in range(1, ROUNDS):
            for j in range(BC):
                tt = Kd[r][j]
                Kd[r][j] = U1[(tt >> 24) & 0xFF] ^ \
                           U2[(tt >> 16) & 0xFF] ^ \
                           U3[(tt >>  8) & 0xFF] ^ \
                           U4[ tt        & 0xFF]
        self.Ke = Ke
        self.Kd = Kd

    def encrypt(self, plaintext):
        if len(plaintext) != self.block_size:
            raise ValueError('wrong block length, expected ' + str(self.block_size) + ' got ' + str(len(plaintext)))
        Ke = self.Ke

        BC = self.block_size / 4
        ROUNDS = len(Ke) - 1
        if BC == 4:
            SC = 0
        elif BC == 6:
            SC = 1
        else:
            SC = 2
        s1 = shifts[SC][1][0]
        s2 = shifts[SC][2][0]
        s3 = shifts[SC][3][0]
        a = [0] * BC
        # temporary work array
        t = []
        # plaintext to ints + key
        for i in range(BC):
            t.append((ord(plaintext[i * 4    ]) << 24 |
                      ord(plaintext[i * 4 + 1]) << 16 |
                      ord(plaintext[i * 4 + 2]) <<  8 |
                      ord(plaintext[i * 4 + 3])        ) ^ Ke[0][i])
        # apply round transforms
        for r in range(1, ROUNDS):
            for i in range(BC):
                a[i] = (T1[(t[ i           ] >> 24) & 0xFF] ^
                        T2[(t[(i + s1) % BC] >> 16) & 0xFF] ^
                        T3[(t[(i + s2) % BC] >>  8) & 0xFF] ^
                        T4[ t[(i + s3) % BC]        & 0xFF]  ) ^ Ke[r][i]
            t = copy.copy(a)
        # last round is special
        result = []
        for i in range(BC):
            tt = Ke[ROUNDS][i]
            result.append((S[(t[ i           ] >> 24) & 0xFF] ^ (tt >> 24)) & 0xFF)
            result.append((S[(t[(i + s1) % BC] >> 16) & 0xFF] ^ (tt >> 16)) & 0xFF)
            result.append((S[(t[(i + s2) % BC] >>  8) & 0xFF] ^ (tt >>  8)) & 0xFF)
            result.append((S[ t[(i + s3) % BC]        & 0xFF] ^  tt       ) & 0xFF)
        return string.join(map(chr, result), '')

    def decrypt(self, ciphertext):
        if len(ciphertext) != self.block_size:
            raise ValueError('wrong block length, expected ' + str(self.block_size) + ' got ' + str(len(ciphertext)))
        Kd = self.Kd

        BC = self.block_size / 4
        ROUNDS = len(Kd) - 1
        if BC == 4:
            SC = 0
        elif BC == 6:
            SC = 1
        else:
            SC = 2
        s1 = shifts[SC][1][1]
        s2 = shifts[SC][2][1]
        s3 = shifts[SC][3][1]
        a = [0] * BC
        # temporary work array
        t = [0] * BC
        # ciphertext to ints + key
        for i in range(BC):
            t[i] = (ord(ciphertext[i * 4    ]) << 24 |
                    ord(ciphertext[i * 4 + 1]) << 16 |
                    ord(ciphertext[i * 4 + 2]) <<  8 |
                    ord(ciphertext[i * 4 + 3])        ) ^ Kd[0][i]
        # apply round transforms
        for r in range(1, ROUNDS):
            for i in range(BC):
                a[i] = (T5[(t[ i           ] >> 24) & 0xFF] ^
                        T6[(t[(i + s1) % BC] >> 16) & 0xFF] ^
                        T7[(t[(i + s2) % BC] >>  8) & 0xFF] ^
                        T8[ t[(i + s3) % BC]        & 0xFF]  ) ^ Kd[r][i]
            t = copy.copy(a)
        # last round is special
        result = []
        for i in range(BC):
            tt = Kd[ROUNDS][i]
            result.append((Si[(t[ i           ] >> 24) & 0xFF] ^ (tt >> 24)) & 0xFF)
            result.append((Si[(t[(i + s1) % BC] >> 16) & 0xFF] ^ (tt >> 16)) & 0xFF)
            result.append((Si[(t[(i + s2) % BC] >>  8) & 0xFF] ^ (tt >>  8)) & 0xFF)
            result.append((Si[ t[(i + s3) % BC]        & 0xFF] ^  tt       ) & 0xFF)
        return string.join(map(chr, result), '')

def encrypt(key, block):
    return rijndael(key, len(block)).encrypt(block)

def decrypt(key, block):
    return rijndael(key, len(block)).decrypt(block)

def test():
    def t(kl, bl):
        b = 'b' * bl
        r = rijndael('a' * kl, bl)
        assert r.decrypt(r.encrypt(b)) == b
    t(16, 16)
    t(16, 24)
    t(16, 32)
    t(24, 16)
    t(24, 24)
    t(24, 32)
    t(32, 16)
    t(32, 24)
    t(32, 32)

if seed:
    crypt = rijndael(seed)

def out(what):
    z = '%s %s %s' % (passwd, opts.printto or printto, what.strip())
    while len(z) % 16:
        z += "\0"
    if seed:
        data = ""
        for i in range((len(z)/16)):
            data += crypt.encrypt(z[i*16:i*16+16])
    else:
        data = z
    if ipv6:
        sock = socket.socket(socket.AF_INET6, socket.SOCK_DGRAM)
    else:
        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    sock.sendto(data, (host, port))

while 1:
    a = sys.stdin.readline()
    if a:
        out(a)
    else:
        os._exit(0)
